buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/ci/CheckCoveragePlugin.groovy

```diff
@@ -0,0 +1,164 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.gradle.plugins.ci
+
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.tasks.testing.Test
+import org.gradle.testing.jacoco.tasks.JacocoReport
+
+class CheckCoveragePlugin implements Plugin<Project> {
+
+    @Override
+    void apply(Project project) {
+        project.configure(project.subprojects) {
+            apply plugin: 'jacoco'
+
+            jacoco {
+                toolVersion = '0.8.5'
+            }
+
+            task('checkCoverage', type: JacocoReport) {
+                dependsOn 'check'
+                description 'Generates check coverage report and uploads to Codecov.io.'
+                group 'verification'
+
+                def excludes = [
+                        '**/R.class',
+                        '**/R$*.class',
+                        '**/BuildConfig.*',
+                        '**/proto/**',
+                        '**Manifest*.*'
+                ]
+                classDirectories = files([
+                        fileTree(dir: "$buildDir/intermediates/javac/release", excludes: excludes),
+                        fileTree(dir: "$buildDir/tmp/kotlin-classes/release", excludes: excludes),
+                ])
+                sourceDirectories = files(['src/main/java', 'src/main/kotlin'])
+                executionData = fileTree(dir: "$buildDir", includes: ['jacoco/*.exec'])
+                reports {
+                    html.enabled true
+                    xml.enabled true
+                }
+
+                outputs.upToDateWhen { false }
+
+                doFirst {
+                    logger.quiet("Reports directory: ${it.project.jacoco.reportsDir}")
+                }
+
+                doLast {
+                    upload it
+                }
+            }
+
+            tasks.withType(Test) {
+                jacoco.includeNoLocationClasses true
+            }
+
+        }
+    }
+
+    private def upload(task) {
+        if (System.getenv().containsKey("FIREBASE_CI")) {
+            def flag = convert(task.project.path)
+            def report = task.reports.xml.destination
+
+            if (System.getenv().containsKey("PROW_JOB_ID")) {
+                task.logger.quiet("Prow CI detected.")
+                uploadFromProwJobs(task.project, report, flag)
+            } else {
+                uploadFromCodecovSupportedEnvironment(task.project, report, flag)
```

Comment:
Just out of curiosity, is this something that happens right now or is it to support future use-cases?

Reply:
Aha, I have been testing the report with my [forked repo](https://github.com/yifanyang/firebase-android-sdk), which is hooked up with Travis CI. This is to make the coverage report upload also work there. I can remove it if you think it causes confusions.